Add models for ion exchange unit

esrawli opened the pull request:
## Addresses Issue: 


## Summary/Motivation:
This is the first PR in a series of PRs adding the ion exchange models. Note that this PR does not include documentation but future PRs will incorporate it. 

## Changes proposed in this PR:
- Ion exchange unit models, example, and tests.

## Reviewer's checklist / merge requirements:
- [x] The head branch (i.e. the "source" of the changes) is not the `main` branch on the PR author's fork
- [x] Documentation
- [x] Tests
- [x] Diagnostic tests for models

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md file
   at the top level of this directory.

2. I represent I am authorized to make the contributions and grant the license. If my employer has
   rights to intellectual property that includes these contributions, I represent that I have
   received permission to make contributions and grant the required license on behalf of that
   employer.

esrawli committed:
Add models for ion exchange unit

esrawli committed:
Correct typos

esrawli committed:
Remove unused libraries

esrawli committed:
Remove Pylint check for invalid name convention

esrawli committed:
Remove empty list, as an argument, in function

blnicho committed:
Update file location in test

blnicho committed:
Rearrange imports to fix linter warning

blnicho committed:
Skip IX tests when watertap is not available

blnicho committed:
disable pylint check for "wrong-import-position" in IX test file

esrawli committed:
Update parameters for pressure drop and bed expansion for S950 resin

esrawli committed:
Update values

esrawli committed:
Update models to correct regeneration issues when extending coverage in test

These models include the following changes:

1. Add new and replace old regeneration terms based on most updated
version of IX model.
2. Add domain to all variables.
3. Keep the IX model as an cation exchange type and comment other
types. Add a configuration error message to inform this to the user.
4. Comment unusued code and added notes to explain why they are commented.
5. Convert bv and N_Sc from Variable/Constraint pairs to Expressions.
6. Reorganize variables.
7. Improve notes.

esrawli committed:
Update example to include a new argument and add bounds to avoid structural errors

esrawli committed:
Add tests to increase coverage

esrawli committed:
Fix typo

sufikaur committed:
Merge branch 'main' into ix-models

agarciadiego commented on src/prommis/ion_exchange/ion_exchange_costing.py:
@bpaul4 can you help review only this file?

bpaul4 replied:
Thanks for adding this model.

Could we format this cost model as was done for the [diafiltration](https://github.com/prommis/prommis/tree/main/src/prommis/nanofiltration/costing) model? WaterTAP-specific global parameters are defined in the "cost block", and the "cost model" is formatted as a class that is compatible with the PrOMMiS costing framework.

Here's a simpler [example](https://github.com/prommis/prommis/blob/main/src/prommis/uky/costing/custom_costing_example.py) of a custom costing model, although it doesn't use WaterTAP parameters.

Another comment, we could have a single method `cost_ion_exchange` or we could split components into their own costing methods - see how the diafiltration model costs each component of the membrane separately. Some components have capital costs and no operating costs, or vice versa, so that is completely supported if you want to break up your components that way.

esrawli committed:
Add documentation for ion exchange unit model class

esrawli committed:
Remove definitions

esrawli committed:
Correct file name

esrawli committed:
Add new costing files for ion exchange unit model

esrawli committed:
Remove commented lines

esrawli committed:
Remove regenerants that are not used for now

esrawli committed:
Call new costing model files and update test values

esrawli committed:
Remove original costing file

esrawli committed:
Set scaling factors for relevant variables and add missing component from IDAES

esrawli committed:
Add test for configuration error message and for table generation

esrawli committed:
Remove unnecessary function

esrawli committed:
Add test for example file

esrawli committed:
Change paths to absolute paths

esrawli committed:
Remove table number

esrawli committed:
Add documentation to base model

esrawli committed:
Run black

esrawli committed:
Add and remove arguments from functions

esrawli committed:
Re-write test to read fuctions from example script

esrawli committed:
Run black

esrawli committed:
Remove unused imports

esrawli committed:
Replace raise error with ConfigurationError option from IDAES

esrawli committed:
Remove unused arguments

esrawli committed:
Update arguments in functions from example and add test for cost model configuration error for regenerant

esrawli committed:
Correct text and run black

esrawli committed:
Merge branch 'main' into ix-models

bpaul4 commented on src/prommis/ion_exchange/costing/ion_exchange_cost_model.py:
```diff
@@ -0,0 +1,695 @@
+    )
+
+    costing = blk.costing_package
+    costing.register_flow_type(regenerant, blk.cost / blk.purity)
```
Reverse the error check

```
if regenerant not in ["NaCl",]:
    raise ConfigurationError(f"Unsupported regenerant type: {regenerant}")
```
I like using lists for string labels because it supports adding additional options later.
Then, initialize the parameters with the values and units so you don't need lines 51-54.

esrawli replied:
I left it this way since we will be including a list with other regenerants in future iterations of the model. That way, we only have to add the values to the dictionary without having to modify the other lines.

bpaul4 reviewed (commented):
@esrawli I left a couple comments related to costing.

Is there a test running this class as a custom model through the standard PrOMMiS costing? If not, that would be useful to ensure compatibility, see the diafiltration test in https://github.com/prommis/prommis/blob/833f5546d70e76ebaf035a6836b6cc97b7e6065a/src/prommis/uky/costing/tests/test_ree_costing.py#L2397 for reference on how to do so.

esrawli committed:
Improve initial values

esrawli committed:
Remove unused call to electricity flow from costing model and remove its call from example and test. Also, include new test to check compatibility with PrOMMiS costing package and create new function in test to build the IX model to avoid repeating the code multiple times

esrawli commented on src/prommis/ion_exchange/costing/ion_exchange_cost_model.py:
```diff
@@ -0,0 +1,695 @@
+
+            # [ESR NOTE: Original costing package had this as a
+            # variable but I changed it to be a parameter.]
+            resin_type = "typeC"
```
Yes, the will.

esrawli committed:
Remove unused components

blnicho committed:
Run black

esrawli commented:
@bpaul4 , I addressed all your comments, including the addition of a new test to check costing compatibility.

bpaul4 commented on src/prommis/ion_exchange/tests/test_ion_exchange_multicomponent.py:
```diff
@@ -758,54 +727,203 @@ def test_config_error_regen_in_costing():
+        capital_loan_repayment_period=10,
+        debt_percentage_of_capex=50,
+        operating_inflation_percentage=3,
+        revenue_inflation_percentage=3,
```
Otherwise, everything for the costing looks good 👍

esrawli replied:
I removed the unused arguments. Thanks for the feedback!

dallan-keylogic commented on src/prommis/ion_exchange/costing/ion_exchange_cost_model.py:
```diff
@@ -0,0 +1,693 @@
+        # Define coefficients for the ion exchange pressure vessels,
+        # backwash/rinse tank, and regeneration solution tank costed
+        # with power equation. Consider col_vol and tank_vol in
+        # gallons. [ESR NOTE: Theoriginal terms were Var() but now I
```
```suggestion
        # gallons. [ESR NOTE: The original terms were Var() but now I
```

dallan-keylogic commented on src/prommis/ion_exchange/costing/ion_exchange_cost_model.py:
```diff
@@ -0,0 +1,693 @@
+        # aggregate costs in function aggregate_costs()
+        @blk.Expression()
+        def direct_capital_cost(b):
+            return blk.capital_cost / blk.cost_factor
```
It would be better to construct the expression using the local variable `b` instead of the variable `blk` that is inherited from the parent scope.

dallan-keylogic commented on src/prommis/ion_exchange/costing/ion_exchange_cost_model.py:
```diff
@@ -0,0 +1,693 @@
+        )
+        bed_vol_ft3 = pyo.units.convert(
+            blk.unit_model.bed_volume, to_units=pyo.units.ft**3
+        )
```
It would be better to either make these named expressions or to move them inside of the constraints/expressions in which they are used.

dallan-keylogic reviewed (commented):
I've got some code style feedback, but my major question is "How does this model differ from the WaterTap model? Why aren't we inheriting from it?"  I see that there are some model structure differences, but it's not clear to me how large those differences are. Could we create an `ion_exchange_base` class that both the WaterTap and PrOMMiS model inherits from?

esrawli committed:
Move capital recovery factor expression from parent class to costing model, set all parameters with units to mutable=true, and fix typo

esrawli committed:
Construct constraints and expressions using the local variable `b` instead of the variable `blk` and move bed_mass_ton inside the constraint where it is used

esrawli committed:
Convert terms to Pyomo expressions

esrawli committed:
Use full names for terms in the IX configuration and update these names in the models that require them

esrawli committed:
Remove unused arguments for NPV

blnicho committed:
Run black

dallan-keylogic commented on src/prommis/ion_exchange/tests/test_ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,918 @@
+    pyo.assert_optimal_termination(scaled_results)
+
+    # Propagate the solution back to the original model
+    init_scaling.propagate_solution(scaled_model, m_nacl)
```
Same deal here.

dallan-keylogic reviewed (commented):
There's still room for improvement here.

I went through and started suggesting variable names that are either more descriptive or in line with IDAES naming conventions, but I hit a wall midway through because I couldn't figure out what the variables were actually supposed to be in a reasonable amount of time.

Additionally, I suggested changes to avoid hardcoding the time point and phase name used in the unit model, as well as suggesting that we use `ipopt_v2` through IDAES's `get_solver`.

dallan-keylogic reviewed (commented):
There's still room for improvement here.

I went through and started suggesting variable names that are either more descriptive or in line with IDAES naming conventions, but I hit a wall midway through because I couldn't figure out what the variables were actually supposed to be in a reasonable amount of time.

Additionally, I suggested changes to avoid hardcoding the time point and phase name used in the unit model, as well as suggesting that we use `ipopt_v2` through IDAES's `get_solver`.

esrawli committed:
Address comments in documentation

esrawli commented on src/prommis/ion_exchange/ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,725 @@
+highlight opportunities for further refinement and extension of the
+model to improve its accuracy and applicability.
+
+1. **Steady-State Approximation**: The IXMC model uses a steady-state continuous approximation to represent the ion exchange batch process. This simplification may not fully capture transient dynamics or time-dependent behavior observed in real-world ion exchange operations.
```
I made the suggested change to the text.

esrawli committed:
Remove commented code and update text in configuration error message

esrawli committed:
Replace first time point 0 with t0 and import ListOf from Pyomo

esrawli committed:
Move configuration error check and update indentation

esrawli commented on src/prommis/ion_exchange/tests/test_ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,918 @@
+            "hazardous_waste": hazardous_waste,
+        }
+
+        m.fs.unit = ix = IonExchangeMultiComp(**ix_config)
```
I applied this change.

esrawli commented on src/prommis/ion_exchange/ion_exchange_multicomponent_base.py:
```diff
@@ -0,0 +1,1362 @@
+
+    def build(self):
+        super().build()
+
```
I included `t0` in model and added the configuration error.

esrawli committed:
Rename configuration num_traps to number_of_trapezoids in the configuration and within the unit models

esrawli committed:
Replace `Liq` phase with indicator p0 declared using the unit phase_list. Also, replace missing time 0 with t0

esrawli commented on src/prommis/ion_exchange/ion_exchange_multicomponent_base.py:
```diff
@@ -0,0 +1,1362 @@
+
+        @self.Constraint(doc="Reynolds number")
+        def eq_Re(b):  # Eq. 3.358, ref[5] Inglezakis + Poulopoulos
+            return b.N_Re * prop_in.visc_k_phase["Liq"] == (
```
I replaced `Liq` with `p0`.

esrawli commented on src/prommis/ion_exchange/ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,725 @@
+            self.reactive_ion_set, doc="Total mass of ion removed from the liquid phase"
+        )
+        def total_mass_removed(b, j):
+            return (1 - b.c_norm_avg[j]) * prop_in.get_material_flow_terms("Liq", j)
```
I replaced `Liq` with `p0`.

esrawli committed:
Remove commented lines and replace 0 value for time with t0

esrawli committed:
Rename Expression with the concentration at breakthrough following the IDAES notation

esrawli committed:
Turn c_trap_min into a param and remove unused param eps

esrawli committed:
Move configuration arguments related to trapezoidal rule to the model that defines the trapezoidal rule equations

esrawli committed:
Remove commented code for trapezoidal rule

esrawli committed:
Rename variables with more descriptive and self explanatory terms while following IDAES properties notation

blnicho committed:
Merge branch 'main' into ix-models

esrawli committed:
Use ipopt_v2 solver through IDAES get_solver function instead of ipopt from WaterTAP

esrawli commented on src/prommis/ion_exchange/ix_freundlich_multicomponent_example.py:
```diff
@@ -0,0 +1,817 @@
+from idaes.core.util.exceptions import ConfigurationError
+
+# Import WaterTAP models and libraries
+from watertap.core.solvers import get_solver
```
Yes, I changed the solver to `ipopt_v2` in the example and the test.

esrawli committed:
Re-write model documentation to match the existent documentation of other unit models in the PrOMMiS repository

blnicho committed:
Run black

blnicho committed:
Fix docstring formatting

blnicho committed:
Fix docstring formatting

esrawli committed:
Correct formatting in tables

esrawli commented on src/prommis/ion_exchange/ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,725 @@
+multicomponent ion exchange processes while building upon the basic
+assumptions and methodologies of the original IX model. Refer to the
+`WaterTAP IX documentation <https://watertap.readthedocs.io/en/stable/technical_reference/unit_models/ion_exchange_0D.html>`_ for more information on the original IX model, including its
+assumptions, equations, and implementation.
```
I re-wrote the documentation piece so this does not apply as it did in original text.

esrawli committed:
Change style of variables in the degrees of freedom section in the documentation

esrawli commented:
> There's still room for improvement here.
> 
> I went through and started suggesting variable names that are either more descriptive or in line with IDAES naming conventions, but I hit a wall midway through because I couldn't figure out what the variables were actually supposed to be in a reasonable amount of time.
> 
> Additionally, I suggested changes to avoid hardcoding the time point and phase name used in the unit model, as well as suggesting that we use `ipopt_v2` through IDAES's `get_solver`.

I addressed most of the comments. The main changes are:
- I renamed variables to have a more descriptive name while trying to follow the IDAES naming conventions.
- I replaced hardcoded time point 0 and phase name.
- I use `ipopt_v2` through IDAES `get_solver` instead of WaterTAP's `get_solver`.
- Applied other recommended changes throughout the models.

esrawli commented on src/prommis/ion_exchange/ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,725 @@
+3. **Data Implementation and Example Construction Step**:
+
+   - Integrate all calculated and known parameters into the IXMC model
+     to solve the system.
```
Text not included in the documentation anymore.

esrawli committed:
Add configuration errors and their tests when time is not equal to 0 and reactive ion set with no target component.

esrawli commented on src/prommis/ion_exchange/ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,725 @@
+        target_component = self.config.target_component
+        reactive_ions = self.config.reactive_ions
+
+        assert target_component in reactive_ions
```
I added the recommended configuration error.

esrawli committed:
Run pylint

esrawli committed:
Fix warning about dangerous default values

esrawli committed:
Improve scaling, boudsn, and initial values for some variables and add tests to check for numerical warnings

esrawli committed:
Remove unused IDAES components

esrawli committed:
Merge branch 'prommis:main' into ix-models

esrawli commented on src/prommis/ion_exchange/ion_exchange_multicomponent_base.py:
```diff
@@ -0,0 +1,1362 @@
+        )
+
+        prop_in = self.process_flow.properties_in[0]
+
```
The property package we are using only supports the liquid phase, so even before getting into the unit model, the property package will throw an error. I left this out for now while testing if there are other flags that will appear before the unit model is touched.

esrawli commented on src/prommis/ion_exchange/tests/test_ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,918 @@
+
+    for key, expected_value in ix_cost_results.items():
+        mv = getattr(ix_cost, key)
+        assert pyo.value(mv) == pytest.approx(expected_value, rel=1e-4)
```
I added new tests for numerical warnings and condition number, where requested.

esrawli commented:
> > There's still room for improvement here.
> > I went through and started suggesting variable names that are either more descriptive or in line with IDAES naming conventions, but I hit a wall midway through because I couldn't figure out what the variables were actually supposed to be in a reasonable amount of time.
> > Additionally, I suggested changes to avoid hardcoding the time point and phase name used in the unit model, as well as suggesting that we use `ipopt_v2` through IDAES's `get_solver`.
> 
> I addressed most of the comments. The main changes are:
> 
>     * I renamed variables to have a more descriptive name while trying to follow the IDAES naming conventions.
> 
>     * I replaced hardcoded time point 0 and phase name.
> 
>     * I use `ipopt_v2` through IDAES `get_solver` instead of WaterTAP's `get_solver`.
> 
>     * Applied other recommended changes throughout the models.

(UPDATE): I finalized addressing all the comments. Thanks for all the recommended changes.

esrawli commented:
@blnicho , this is ready for your review. Thanks!

esrawli commented on src/prommis/ion_exchange/ion_exchange_multicomponent_base.py:
```diff
@@ -0,0 +1,1362 @@
+                + b.bed_expansion_frac_param_C * b.backwash_loading_rate**2
+            )
+
+    def initialize_build(
+        self,
+        state_args=None,
+        outlvl=idaeslog.NOTSET,
+        solver=None,
+        optarg=None,
+    ):
+        """
```
I think the block triangularization is good enough to initialize this model.

esrawli committed:
Merge branch 'prommis:main' into ix-models

esrawli committed:
Update scaling and test when regeneration is used

ksbeattie committed:
Merge branch 'main' into ix-models

dallan-keylogic reviewed (approved):
LGTM

dallan-keylogic commented:
```
_____________________ test_scaling_and_numerical_warnings _____________________

m = <pyomo.core.base.PyomoModel.ConcreteModel object at 0x000001C2AB20AE30>

    @pytest.mark.unit
    def test_scaling_and_numerical_warnings(m):
    
        m.fs.unit_ix.calculate_scaling_factors()
    
        dt = DiagnosticsToolbox(m)
        dt.assert_no_numerical_warnings(ignore_parallel_components=True)
    
>       assert jacobian_cond(m, scaled=False) == pytest.approx(170184365219283.22)
E       assert np.float64(170182328233489.38) == 170184365219283.22 � 1.7e+08
E         comparison failed
E         Obtained: 170182328233489.38
E         Expected: 170184365219283.22 � 1.7e+08

src\prommis\ion_exchange\tests\test_ion_exchange_multicomponent.py:510: AssertionError
```

You can reduce the relative tolerance to `1e-2` here. We just want an order-of-magnitude estimate of the unscaled condition number to compare to the scaled condition number.

blnicho commented on src/prommis/ion_exchange/costing/ion_exchange_cost_block.py:
```diff
@@ -0,0 +1,98 @@
+#####################################################################################################
+# “PrOMMiS” was produced under the DOE Process Optimization and Modeling for Minerals Sustainability
+# (“PrOMMiS”) initiative, and is copyright (c) 2023-2025 by the software owners: The Regents of the
```
```suggestion
# (“PrOMMiS”) initiative, and is copyright (c) 2023-2026 by the software owners: The Regents of the
```

blnicho commented on src/prommis/ion_exchange/costing/ion_exchange_cost_block.py:
```diff
@@ -0,0 +1,98 @@
+# and
+# https://github.com/watertap-org/watertap/blob/723576a4a05596cc1f897376863c965f22efa99f/watertap/costing/util.py
+
+# WaterTAP Copyright (c) 2020-2025, The Regents of the University of California,
```
```suggestion
# WaterTAP Copyright (c) 2020-2026, The Regents of the University of California,
```

blnicho committed:
Relax test tolerance for checking IX condition number

blnicho committed:
Adding missing __init__.py file

blnicho commented on src/prommis/ion_exchange/costing/ion_exchange_cost_model.py:
```diff
@@ -0,0 +1,710 @@
+#####################################################################################################
+# “PrOMMiS” was produced under the DOE Process Optimization and Modeling for Minerals Sustainability
+# (“PrOMMiS”) initiative, and is copyright (c) 2023-2025 by the software owners: The Regents of the
```
```suggestion
# (“PrOMMiS”) initiative, and is copyright (c) 2023-2026 by the software owners: The Regents of the
```

blnicho commented on src/prommis/ion_exchange/costing/ion_exchange_cost_model.py:
```diff
@@ -0,0 +1,710 @@
+# and
+# https://github.com/watertap-org/watertap/blob/main/watertap/costing/watertap_costing_package.py
+
+# WaterTAP Copyright (c) 2020-2025, The Regents of the University of California,
```
```suggestion
# WaterTAP Copyright (c) 2020-2026, The Regents of the University of California,
```

blnicho commented on src/prommis/ion_exchange/tests/test_ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,1024 @@
+#####################################################################################################
+# “PrOMMiS” was produced under the DOE Process Optimization and Modeling for Minerals Sustainability
+# (“PrOMMiS”) initiative, and is copyright (c) 2023-2025 by the software owners: The Regents of the
```
```suggestion
# (“PrOMMiS”) initiative, and is copyright (c) 2023-2026 by the software owners: The Regents of the
```

blnicho commented on src/prommis/ion_exchange/tests/test_ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,1024 @@
+
+# WaterTAP License Agreement
+
+# WaterTAP Copyright (c) 2020-2025, The Regents of the University of California,
```
```suggestion
# WaterTAP Copyright (c) 2020-2026, The Regents of the University of California,
```

blnicho commented on src/prommis/ion_exchange/ix_freundlich_multicomponent_example.py:
```diff
@@ -0,0 +1,837 @@
+#####################################################################################################
+# “PrOMMiS” was produced under the DOE Process Optimization and Modeling for Minerals Sustainability
+# (“PrOMMiS”) initiative, and is copyright (c) 2023-2025 by the software owners: The Regents of the
```
```suggestion
# (“PrOMMiS”) initiative, and is copyright (c) 2023-2026 by the software owners: The Regents of the
```

blnicho commented on src/prommis/ion_exchange/ix_freundlich_multicomponent_example.py:
```diff
@@ -0,0 +1,837 @@
+
+# WaterTAP License Agreement
+
+# WaterTAP Copyright (c) 2020-2025, The Regents of the University of California,
```
```suggestion
# WaterTAP Copyright (c) 2020-2026, The Regents of the University of California,
```

blnicho commented on src/prommis/ion_exchange/ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,731 @@
+#####################################################################################################
+# “PrOMMiS” was produced under the DOE Process Optimization and Modeling for Minerals Sustainability
+# (“PrOMMiS”) initiative, and is copyright (c) 2023-2025 by the software owners: The Regents of the
```
```suggestion
# (“PrOMMiS”) initiative, and is copyright (c) 2023-2026 by the software owners: The Regents of the
```

blnicho commented on src/prommis/ion_exchange/ion_exchange_multicomponent.py:
```diff
@@ -0,0 +1,731 @@
+# This model was derived from:
+# https://github.com/kurbansitterley/watertap/blob/ix_reorg/watertap/unit_models/ion_exchange/ion_exchange_multi_component.py
+
+# WaterTAP Copyright (c) 2020-2025, The Regents of the University of California,
```
```suggestion
# WaterTAP Copyright (c) 2020-2026, The Regents of the University of California,
```

blnicho commented on src/prommis/ion_exchange/ion_exchange_multicomponent_base.py:
```diff
@@ -0,0 +1,1138 @@
+#####################################################################################################
+# “PrOMMiS” was produced under the DOE Process Optimization and Modeling for Minerals Sustainability
+# (“PrOMMiS”) initiative, and is copyright (c) 2023-2025 by the software owners: The Regents of the
```
```suggestion
# (“PrOMMiS”) initiative, and is copyright (c) 2023-2026 by the software owners: The Regents of the
```

blnicho commented on src/prommis/ion_exchange/ion_exchange_multicomponent_base.py:
```diff
@@ -0,0 +1,1138 @@
+
+# WaterTAP License Agreement
+
+# WaterTAP Copyright (c) 2020-2024, The Regents of the University of California,
```
```suggestion
# WaterTAP Copyright (c) 2020-2026, The Regents of the University of California,
```

blnicho committed:
Update dates in copyright statements

Co-authored-by: Bethany Nicholson <blnicho@users.noreply.github.com>

blnicho committed:
Remove unused imports
